### PR TITLE
Merging Set up of Base NPC

### PR DESCRIPTION
**Pull request by NGrinstead0101**
Created a script with baseline NPC logic that other unique NPCs can inherit from. Also made a SampleNpc script and prefab for testing state transitions and cycling through dialogue.

### CODE REVIEW

File: Assets/Scripts/NpcBehaviors/SampleNpc.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+*    Description: Sample class showing how an NPC can be created from
+*       BaseNPC. Can also be used for testing purposes.
+*******************************************************************/
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class SampleNpc : BaseNpc
+{
+    private void Awake()
```
Comment (alecpizz) [inline Assets/Scripts/NpcBehaviors/SampleNpc.cs:14]: Why not have the baseNPC handle invoke this?

File: Assets/Scripts/NpcBehaviors/BaseNpc.cs [hunk 2]
```diff
@@ -82,34 +83,77 @@ protected struct DialogueGroup
...
     public void Interact()
     {
         if (_canInteract)
         {
-            // TODO: display dialogue based on current state
+            DialogueGroup currentDialogue = _stateDialogue.GetStateData(_currentState);
+            
+            if (currentDialogue.CanTalk)
+            {
+                if (_isInteracting == false)
```
Comment (alecpizz) [inline Assets/Scripts/NpcBehaviors/BaseNpc.cs:124]: !_isInteracting

File: Assets/Scripts/NpcBehaviors/BaseNpc.cs [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+{
+    /// <summary>
+    /// Structure used to store a collection of data related to the different states
+    /// Example: A StateDataGroup<Animation> could hold idle animations for each state
+    /// </summary>
+    /// <typeparam name="T">Type of data to be stored</typeparam>
+    [System.Serializable]
+    protected struct StateDataGroup<T>
+    {
+        [SerializeField] T _idleState;
```
Comment (alecpizz) [inline Assets/Scripts/NpcBehaviors/BaseNpc.cs:25]: I know that these are inherently private by default, do we need to be more explicit? No idea what the standard requires, but consider adding a private.

File: Assets/Scripts/NpcBehaviors/BaseNpc.cs [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+public abstract class BaseNpc : MonoBehaviour
+{
+    /// <summary>
+    /// Structure used to store a collection of data related to the different states
+    /// Example: A StateDataGroup<Animation> could hold idle animations for each state
+    /// </summary>
+    /// <typeparam name="T">Type of data to be stored</typeparam>
+    [System.Serializable]
+    protected struct StateDataGroup<T>
```
Comment (alecpizz) [inline Assets/Scripts/NpcBehaviors/BaseNpc.cs:23]: I really like this approach to handling NPC data.

File: Assets/Scripts/NpcBehaviors/BaseNpc.cs [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    [SerializeField] protected StateDataGroup<DialogueGroup> _stateDialogue;
+    [SerializeField] protected StateDataGroup<Vector3> _navigationPositions;
+    [SerializeField] protected StateDataGroup<Animation> _stateAnimations;
+
+    protected PlayerControls _playerControls;
+    protected NavMeshAgent _navAgent;
+    protected Animator _animator;
+
+    protected NpcStates _currentState = NpcStates.DefaultIdle;
+    protected bool[] _canChangeStates = { false, false, false, false, false };
```
Comment (alecpizz) [inline Assets/Scripts/NpcBehaviors/BaseNpc.cs:91]: What is this being used for?

**Comment by NGrinstead0101** [comment]
Made some changes based on your feedback. I took the _canChangeStates variable out since it was an old idea that I just forgot to remove.

**Pull request merged**